### ISSUE TITLE
More advanced JSON to custom structure examples

### ISSUE THREAD

**Issue by wnewbery** [miloyip/rapidjson#671]
I want to parse some JSON structures into C++ objects, and am trying to determine the best approach. There is an example ( http://rapidjson.org/md_doc_sax.html#CustomDataStructure ) but that is only an extremely simple case and using a handcrafted enum like that quickly becomes very complex.

Using the DOM model, something along the lines of `void read_json(const rapidjson::Value &val, SomeType *out)` with various overloads and helpers is possible (and a `write_json(writer,in)` along similar lines), but would be very interested if anyone has already worked out such a design for the SAX handler API.

**Comment by mbroadst** [comment]
I would be very interested in this as well, currently I think I'm running into some memory issues because I'm parsing a document from a buffer in situ and trying to copy out the values - I can't find any examples out there of parsing this way. Does anyone have a simple example along these lines?

**Comment by miloyip** [comment]
Please check if #1138 is useful for you. 
Comments are welcome.

**Comment by wnewbery** [comment]
Sorry if I misread the code, but that appears to also be parsing the entire JSON document into a DOM first (`JsonReader::mDocument`), then walking over that, rather than parsing it straight into the objects?

**Comment by miloyip** [comment]
Yes. Using SAX to do deserialization is more difficult. But the major feature of that example is to using a (mostly) symmetric way to serialize/deserialize object. You may still try to make a framework which uses SAX to do so.

**Comment by carolinebeltran** [comment]
Using DOM, it is not difficult to parse a **'known' JSON structure** into C++ objects, such as a std container that contains multiple child containers.  What I cannot figure out is how to parse an **'unknown' JSON structure**.  I am trying to do this using SAX instead of DOM but it is more complex than I thought.  In a nutshell, I want to simply **pass any valid JSON string** to a function and have the function return the **C++ object** (map of other maps and vectors) that contains all of the parsed JSON data.  To store any type of value such as string, int, bool, double, I am using Boost Variant.  Please advise.

**Comment by miloyip** [comment]
@carolinebeltran
> I am trying to do this using SAX instead of DOM but it is more complex than I thought. In a nutshell, I want to simply pass any valid JSON string to a function and have the function return the C++ object (map of other maps and vectors) that contains all of the parsed JSON data. 

The `rapidjson::Document` is a C++ object that stores arbitrary JSON. The `rapidjson::Value` is a variant type.

**Comment by Woodsigel** [comment]
Hi @wnewbery  @carolinebeltran . Here is my solution built on RapidJSON — it should address your requirements. [RapidJSON-utility](https://github.com/Woodsigel/RapidJSON-utility)
